test/helpers/kubectl.go

```diff
@@ -4102,3 +4102,30 @@ func (kub *Kubectl) GetDNSProxyPort(ciliumPod string) int {
 	}
 	return port
 }
+
+// AddIPRoute adds a route to a given subnet address and a gateway on a given
+// node via the iproute2 utility suite. The function takes in a flag called
+// replace which will convert the action to replace the  route being added if
+// another route exists and matches. This allows for idempotency as the "replace"
+// action will not fail if another matching route exists, whereas "add" will fail.
+func (kub *Kubectl) AddIPRoute(nodeName, subnet, gw string, replace bool) {
+	action := "add"
+	if replace {
+		action = "replace"
+	}
+	cmd := fmt.Sprintf("ip route %s %s via %s", action, subnet, gw)
+
+	res, err := kub.ExecInHostNetNS(context.TODO(), nodeName, cmd)
+	gomega.Expect(res).To(CMDSuccess())
+	gomega.Expect(err).ToNot(gomega.HaveOccurred(), "Cannot exec %s in node %s", cmd, nodeName)
```

Comment:
I think this refactor is a good one overall. 

The issue I have with this is what I've highlighted. In general, we should avoid doing expectations inside helper functions. The rationale is that it screws up the stacktrace when the test fails. As an example here, if `AddIPRoute` fails, the test output will point at this funciton, but not *where* `AddIPRoute` was called. To workaround that, we can use `ExpectWithOffset`. However, I think we should avoid using the offset variation when possible, to simplify thinking about the case when there's potentially nested calls. (I've also seen variations of our helpers accept an `offset` to pass into `ExpectWithOffset`, but I feel this is a bit too much. We should strive for simplicity instead.)

I would propose to return the `res` and `err` instead, and let the caller of these helpers perform the expectations. I understand trying to reduce the boilerplate of doing expectations on helper functions, but I think having consistent and readable stacktraces when a test fails is more valuable than "elegant" test code, if you will :).

Reply:
Yeah, good points re the offset. Anyway, please take a look at another approach (two last commits).